src/binary/raw_binary_reader.rs

```diff
@@ -1049,16 +1098,25 @@ where
         Ok(field_id)
     }
 
-    fn read_annotations(&mut self) -> IonResult<()> {
+    /// Reads the annotations in the annotations wrapper and returns the expected length of the
+    /// wrapped value that will follow.
+    fn read_annotations(&mut self) -> IonResult<usize> {
         let num_annotations_before = self.cursor.annotations.len();
         // The first byte of the annotations envelope is now behind the cursor
         let annotations_offset = self.cursor.bytes_read - 1;
-        // The encoding allows us to skip over the annotations list and the value, but in practice
-        // we won't know if we want to skip this value until we've read the type descriptor byte.
-        // That means we need to read the length even though we have no intent to use it.
-        let _annotations_and_value_length = self.read_standard_length()?;
+        let annotations_and_value_length = self.read_standard_length()?;
```

Comment:
🗺️ The total annotations wrapper length (`annotations` length + `value` length) is often redundant for the 'happy path'. However, it's useful for detecting malformed streams. We now use it to confirm that the wrapper and the wrapped value's lengths agree.